com.unity.ml-agents/Runtime/InplaceArray.cs

```diff
@@ -0,0 +1,250 @@
+using System;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents
+{
+    /// <summary>
+    /// An array-like object that stores up to four elements.
+    /// This is a value type that does not allocate any additional memory.
+    /// </summary>
+    /// <remarks>
+    /// This does not implement any interfaces such as IList, in order to avoid any accidental boxing allocations.
```

Comment:
Small tradeoff between ergonomics and efficiency. It's small either way, so I'm willing to change this.